### PR TITLE
Add support for Kobo Libra H2O

### LINKED ISSUE #882: Kobo Libra H2O touch rotated and partly unreachable
XCSoar 7.23
Kobo Libra H2O
-------------------------------------------------

Expected behavior
-----------------

Touch screen is usable.

Actual behavior
---------------

Touch input seems 90° rotated. Button shows animation for action despite touching the 90° rotated position.
Not the whole screen is reachable.

Steps to reproduce the behavior
-------------------------------

Install XCSoar 7.23 on Kobo Libra H2O

### PR DESCRIPTION
**Pull request by cmoetzing**
The device has similar hardware as Kobo Clara HD.

Related issues and discussions
------------------------------
Closes #882

### CODE REVIEW

File: src/Kobo/System.cpp [hunk 1]
```diff
@@ -178,7 +179,7 @@ void
 KoboUnexportUSBStorage()
 {
 #ifdef KOBO
-  if(DetectKoboModel() == KoboModel::CLARA_HD)
+  if(DetectKoboModel() == KoboModel::CLARA_HD || DetectKoboModel() == KoboModel::LIBRA_H2O)
```
Comment (MaxKellermann) [inline src/Kobo/System.cpp:182]: Line toooooooooooooooooooooooooo long

Comment (cmoetzing) [reply src/Kobo/System.cpp]: add a line break

File: src/DisplayOrientation.hpp [hunk 2]
```diff
@@ -47,9 +49,13 @@ static constexpr DisplayOrientation DEFAULT_DISPLAY_ORIENTATION =
 constexpr DisplayOrientation
 TranslateDefaultDisplayOrientation(DisplayOrientation orientation)
 {
-  return orientation == DisplayOrientation::DEFAULT
-    ? DEFAULT_DISPLAY_ORIENTATION
-    : orientation;
+  if(orientation == DisplayOrientation::DEFAULT) {
+    if(DetectKoboModel() == KoboModel::LIBRA_H2O) {
```
[the remark this answers is not in the document]
Comment (cmoetzing) [reply src/DisplayOrientation.hpp:53]: First of all: I am not a C++ developer. :)
I can compile the code and it runs on my Kobo (with correct screen input).
I have no idead what `constexpr` means. Can you please elaborate why the function should not be called here and what a solution would be?

**Comment by lordfolken** [comment]
hmm you removed our workflows. You can disable the workflows in your repository settings.

**Review by lordfolken — CHANGES_REQUESTED** [review]
Checkout your branch before your changes and then do
```git cherry-pick <commitid>```
Then force push again. 
that way you can eliminate this change

**Comment by cmoetzing** [comment]
I don't know if you allow global variables. Overall this is a cleaner version (I think). Open for suggestions.

**Comment by MaxKellermann** [comment]
I'm confused by that commit diff. It moves chunks of code around, but it's documented to add a feature. I can't see what is really changed.

**Comment by cmoetzing** [comment]
> I'm confused by that commit diff. It moves chunks of code around, but it's documented to add a feature. I can't see what is really changed.

Split it into several commits.

File: src/DisplayOrientation.hpp [hunk 3]
```diff
@@ -34,25 +34,18 @@ enum class DisplayOrientation : uint8_t {
...
 };
 
-#ifdef KOBO
-/* Kobo defaults to portrait */
-static constexpr DisplayOrientation DEFAULT_DISPLAY_ORIENTATION =
-  DisplayOrientation::PORTRAIT;
-#else
-/* everything else defaults to landscape */
-static constexpr DisplayOrientation DEFAULT_DISPLAY_ORIENTATION =
+static DisplayOrientation default_display_orientation =
```
Comment (MaxKellermann) [inline src/DisplayOrientation.hpp]: I don't understand this whole commit, and the commit message doesn't explain anything.

File: src/Kobo/Model.cpp [hunk 4]
```diff
@@ -78,12 +82,22 @@ DetectKoboModel(const char *p)
   return KoboModel::UNKNOWN;
 }
 
-KoboModel
-DetectKoboModel()
-{
+void InitKoboModel() {
```
Comment (MaxKellermann) [inline src/Kobo/Model.cpp:85]: Bad coding style

File: src/Kobo/Model.hpp [hunk 5]
```diff
@@ -37,10 +37,11 @@ enum class KoboModel {
...
   NIA,
+  LIBRA_H2O
 };
 
-gcc_const
-KoboModel
-DetectKoboModel();
+void InitKoboModel();
+
+KoboModel GetKoboModel();
```
Comment (MaxKellermann) [inline src/Kobo/Model.hpp:45]: This is an API change which has nothing to do with Libra H2O, does it? This change obfuscates the commitdiff and makes review very hard. I don't even know why you're doing this; your commit message doesn't explain anything.

File: src/Kobo/Model.cpp [hunk 6]
```diff
@@ -78,12 +82,22 @@ DetectKoboModel(const char *p)
...
   char buffer[16];
-  if (!ReadFromFile("/dev/mmcblk0", 0x200, buffer, sizeof(buffer)))
-    return KoboModel::UNKNOWN;
+    if (!ReadFromFile("/dev/mmcblk0", 0x200, buffer, sizeof(buffer)))
```
Comment (MaxKellermann) [inline src/Kobo/Model.cpp:87]: Bad indent, changed for no obvious reason to 4 spaces.

File: src/Kobo/Model.cpp [hunk 7]
```diff
@@ -78,12 +82,22 @@ DetectKoboModel(const char *p)
...
+      kobo_model = KoboModel::UNKNOWN;
+    else
+      kobo_model = DetectKoboModel(buffer);
+
+  switch(kobo_model) {
+    case KoboModel::LIBRA_H2O:
```
Comment (MaxKellermann) [inline src/Kobo/Model.cpp:93]: Bad indent.

File: src/Kobo/PowerOff.cpp [hunk 8]
```diff
@@ -164,7 +168,7 @@ int main(int argc, char **argv)
 
   /* now we can power off the Kobo; the picture remains on the
      screen */
-  if (DetectKoboModel() == KoboModel::GLO_HD)
+  if (GetKoboModel() == KoboModel::GLO_HD)
```
Comment (MaxKellermann) [inline src/Kobo/PowerOff.cpp:171]: Lots of changes unrelated to Libra H2O

**Review by MaxKellermann — CHANGES_REQUESTED** [review]
I'm confused by this PR.

**Comment by MaxKellermann** [comment]
> Input and display does not work properly

What does that mean, exactly? "Does not work" is the worst possible problem description, because all it does is state the presence of the problem, but says nothing else.

>  The Libra uses a different default display orientation.

What does "default display orientation" mean in this context? How is it part of the problem?

> The above is why there are also changes in places general Kobo stuff is handled, not only Kobo Libra stuff.

You can do any number of things in one PR, those different things should be separate commits, or else you make it very hard to review your changes. The second thing that made your PR very hard to review is the lack of a good commit message; there was no explanation for anything.

**Comment by MaxKellermann** [comment]
> With "default display orientation" I mean the literal meaning of the words: the orientation of the drawn Image on the screen if you do a fresh install on the device an power it up the first time without changing any settings. The default drawing direction of XCSoar means I see a 90° rotated image on the device which runs outside the screen edges so not the whole image is visible and touch input does not trigger actions under you finger but approx. 90° rotated aswell.

Where does this come from? Is the touch screen input rotated as well? What is the fb rotation setting?

**Comment by MaxKellermann** [comment]
> How do I check?

`/sys/class/graphics/fb0/rotate` and `/sys/class/graphics/fbcon/rotate`

**Comment by cmoetzing** [comment]
```
~ # cat /sys/class/graphics/fb0/rotate
3
```
File `/sys/class/graphics/fbcon/rotate` does not exist.

**Comment by ssosas** [comment]
Hi all, I have a similar problem on Aura H2O with xcsoar v7.25. All works fine but display is mirrored (180°) and you need to guess where real buttons are to interact with the touchscreen.

Have you any v7.xx beta which fix that?

Many thanks!

**Comment by cmoetzing** [comment]
Base on the recent support for the Libra 2 I updated my PR. It is indeed a lot cleaner now.

**Pull request merged**